src/main/java/net/minecraftforge/common/crafting/NBTIngredientFuzzy.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.crafting;
+
+import java.util.stream.Stream;
+
+import javax.annotation.Nullable;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+import net.minecraft.advancements.critereon.NbtPredicate;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.network.FriendlyByteBuf;
+
+public class NBTIngredientFuzzy extends Ingredient
+{
+    private final ItemStack stack;
+    private final NbtPredicate predicate;
+    public NBTIngredientFuzzy(ItemStack stack)
+    {
+        super(Stream.of(new ItemValue(stack)));
+        this.stack = stack;
+        this.predicate = stack.hasTag() ? new NbtPredicate(stack.getShareTag()) : NbtPredicate.ANY;
+    }
+
+    @Override
+    public boolean test(@Nullable ItemStack input)
+    {
+        if (input == null)
+            return false;
+        return this.stack.getItem() == input.getItem() && this.stack.getDamageValue() == input.getDamageValue() && this.predicate.matches(input);
```

Comment:
I am not sure this even really matters given the partial matching but I think this potentially should end with `this.predicate.matches(input.getShareTag())` so that it is comparing the share tag of the stack similar to how the predicate is built from the share tag? (Though I am not quite sure that matches method then short circuits the `NbtPredicate.ANY` to being true so then it potentially would need that check as well)